code/datums/ai/ai.dm

```diff
@@ -1,5 +1,35 @@
 /datum/ai
 	var/name
+	var/mob/living/body		// The parent mob we control.
+	var/wait_for = 	0 		// The next time we can process.
+	var/run_interval = 1 	// How long to wait between processes.
+	
+/datum/ai/New(var/mob/living/body)
+	src.body = body
+	START_PROCESSING(SSai, src)
 
-/datum/ai/proc/process()
-	return PROCESS_KILL
+/datum/ai/proc/can_process()
+	if((body.client || body.mind) && !(body.status_flags & ENABLE_AI))
+		return FALSE
+	if(wait_for > world.time)
+		return FALSE
+	if(body.stat == DEAD)
+		return FALSE
+	return TRUE
+
+/datum/ai/Process()
+	if(!can_process())
+		return
+
+	var/time_elapsed = wait_for - world.time
+	wait_for = world.time + run_interval
+	do_process(time_elapsed)
+
+// This is the place to actually do work in the AI.
+/datum/ai/proc/do_process(var/time_elapsed)
+	
+
+/datum/ai/Destroy()
+	STOP_PROCESSING(SSai, src)
+	body = null
```

Comment:
`if(body.ai == src) body.ai = null`

Reply:
Why?

Reply:
Because this will not GC while the mob holds a reference to it.

Reply:
It's QDEL_NULL'd in the mob.

Reply:
And if something deletes the datum rather than the mob, it won't GC.